### PR TITLE
added game option to disable unwanted spectators from a multiplayer game

### PR DESCRIPTION
**Pull request by alexban011**
resolves discord suggestion

Added an option to disable peoples ability to enter and "spy" on your game so people can't cheat like that.

When loading the game it checks if your userId is in the list of civInfoPreview and if it isn't it displays a ToastPopup informing you that you're not allowed to watch that game and obviously, doesn't load the game :)
This means that normal spectators can still be chosen even with the option enabled

Video is out of date

https://user-images.githubusercontent.com/21990802/175565896-c2d2c395-69ea-4c61-85b1-b41aa065b89b.mp4

### CODE REVIEW

**Comment by OptimizedForDensity** [comment]
"No uninvited spectators" isn't translatable.

**Comment by alexban011** [comment]
got it

File: core/src/com/unciv/ui/multiplayer/MultiplayerHelpers.kt [hunk 1]
```diff
@@ -30,6 +31,19 @@ object MultiplayerHelpers {
...
+        if (selectedGame.preview?.gameParameters?.noUnwelcomeSpectators == true) {
+            if (selectedGame.preview!!.civilizations.any { it.playerId == UncivGame.Current.settings.multiplayer.userId }) {
+                actuallyLoadGame(selectedGame, loadingGamePopup)
+            } else {
+                loadingGamePopup.close()
+                ToastPopup("You are not allowed to watch this game!", screen).open()
+            }
+        } else {
+            actuallyLoadGame(selectedGame, loadingGamePopup)
+        }
```
[the remark this answers is not in the document]
Comment (alexban011) [reply core/src/com/unciv/ui/multiplayer/MultiplayerHelpers.kt:43]: hmm I can't put it in `UncivGame.loadGame` cause I can't return and have no BaseScreen/Stage to put the popup on. Would it be ok to put the check everywhere where `loadGame` is called?

[the remark this answers is not in the document]
Comment (alexban011) [reply core/src/com/unciv/ui/multiplayer/MultiplayerHelpers.kt:43]: alright, I'll convert this to draft until #7255 is merged

File: android/assets/jsons/translations/template.properties [hunk 2]
```diff
@@ -601,6 +602,7 @@ Paste gameID from clipboard =
 GameID = 
 Game name = 
 Loading latest game state... = 
+You are not allowed to watch this game! =
```
Comment (yairm210) [inline android/assets/jsons/translations/template.properties:605]: Allowed to *spectate*. so our terminology is consistent

File: core/src/com/unciv/UncivGame.kt [hunk 3]
```diff
@@ -173,6 +174,12 @@ class UncivGame(parameters: UncivGameParameters) : Game() {
         val prevGameInfo = gameInfo
         gameInfo = newGameInfo
 
+        if (gameInfo?.gameParameters?.anyoneCanSpectate == false) {
+            if (!gameInfo!!.civilizations.any { it.playerId == settings.multiplayer.userId }) {
```
Comment (yairm210) [inline core/src/com/unciv/UncivGame.kt:178]: !any -> none

File: core/src/com/unciv/ui/newgamescreen/NewGameScreen.kt [hunk 4]
```diff
@@ -293,7 +295,19 @@ class NewGameScreen(
...
+        } catch (ex: Exception) {
+            val (message) = LoadGameScreen.getLoadExceptionMessage(ex)
+            launchOnGLThread {
+                popup.reuseWith(message, true)
+                rightSideButton.enable()
+                rightSideButton.setText("Start game!".tr())
+            }
+            Gdx.input.inputProcessor = stage
+            return@coroutineScope
+        }
```
Comment (yairm210) [inline core/src/com/unciv/ui/newgamescreen/NewGameScreen.kt:310]: Waitasecond... you can create a game that you can't load? That sounds like a mistake.
Maybe if someone tries to create a MP game and his ID isn't in there then warn him?

File: core/src/com/unciv/ui/worldscreen/WorldScreen.kt [hunk 5]
```diff
@@ -829,6 +830,12 @@ private fun startNewScreenJob(gameInfo: GameInfo) {
     Concurrency.run {
         val newWorldScreen = try {
             UncivGame.Current.loadGame(gameInfo)
+        } catch (notAPlayer: UncivShowableException) {
+            withGLContext {
+                val mainMenu = UncivGame.Current.goToMainMenu()
+                ToastPopup("Not enough memory on phone to load game!", mainMenu)
```
Comment (yairm210) [inline core/src/com/unciv/ui/worldscreen/WorldScreen.kt:836]: I assume you meant to take the exception message

**Review by yairm210 — APPROVED** [review]

**Comment by alexban011** [comment]
Just noticed the conflics, I'll resolve them as soon as I find some free time :)

**Pull request merged**